parser-typechecker/src/Unison/Hashing/V2/ABT.hs

```diff
@@ -78,27 +79,25 @@ hash :: forall f v a h . (Functor f, Hashable1 f, Eq v, Show v, Ord h, Accumulat
      => Term f v a -> h
 hash = hash' [] where
   hash' :: [Either [v] v] -> Term f v a -> h
-  hash' env (Term _ _ t) = case t of
-    Var v -> maybe die hashInt ind
+  hash' env = \case
+    Var' v -> maybe die hashInt ind
       where lookup (Left cycle) = v `elem` cycle
             lookup (Right v') = v == v'
             ind = findIndex lookup env
             hashInt :: Int -> h
             hashInt i = Hashable.accumulate [Hashable.Nat $ fromIntegral i]
             die = error $ "unknown var in environment: " ++ show v
                         ++ " environment = " ++ show env
-    Cycle (AbsN' vs t) -> hash' (Left vs : env) t
-    -- Cycle t -> hash' env t
-    Abs v t -> hash' (Right v : env) t
-    Tm t -> Hashable.hash1 (hashCycle env) (hash' env) t
+    Cycle' vs t -> Hashable.hash1 (hashCycle vs env) undefined t
```

Comment:
I think the need to use `undefined` here means we should probably re-visit this typeclass implementation at some point in the future, at a glance I don't understand why `hash1` needs two separate hashing function args, but maybe there's a reason.

Reply:
> at a glance I don't understand why hash1 needs two separate hashing function args, but maybe there's a reason.

There's some docs on `Hashable1` explaining this. Also https://github.com/unisonweb/unison/pull/2777#discussion_r778226077

@mitchellwrosen and I discussed that there might be a valid implementation that could be passed here instead of `undefined` (I think it would be like `snd . hashCycle vs (Left [] : env) []`) but we'd rather complain loudly since we expect that the functor nested immediately under a `Cycle` will only use the `hashCycle` argument for that level. The reason for that is that all the variables introduced by the cycle will be in scope when reaching whatever the underlying term is at that level, so using the non-cycle hashing function is probably a mistake.

The ABT structure is pretty nice, but it doesn't try to encode every invariant in the types, and this is an example of how that makes something a little awkward. There are more typeful ways of encoding the binding structure of a language (like the bound library) but ABTs are a good tradeoff. If you would like we could nerd out about that sometime... probably not on this PR though. :)